design/metadata-handling.md

```diff
@@ -0,0 +1,417 @@
+
+### Risks and Mitigations
+
+From a user point of view, if this is not well documented or understood, it
+could lead to unexpected errors or behaviors. The errors have to be easily
+understandable.
+
+## Design Details
+
+BareMetalHost would have an added field `metaData` in the `Spec` pointing to a
+secret containing the metadata map. Bare Metal Operator would then pass this map
+to Ironic.
+
+The Metal3Machine will be modified as follow:
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+  name: machine-1
+  namespace: default
+spec:
+  dataTemplate:
+    name: nodepool-1
+    namespace: default
+```
+
+or alternatively
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3Machine
+metadata:
+  name: machine-1
+  namespace: default
+spec:
+  metaData:
+    name: machine-1-metadata
+    namespace: default
+  networkData:
+    name: machine-1-networkData
+    namespace: default
+```
+
+A `metaData` field would be added, consisting of a secret reference containing
+the name and the namespace of the secret containing the metaData for this
+Metal3Machine. A `networkData` field would be added, consisting of a secret
+reference containing the name and the namespace of the secret containing the
+network configuration for this Metal3Machine.
+
+A `dataTemplate` field would be added, consisting of an object reference
+containing the name and the namespace of the secret containing the templates for
+the metadata and network data generation for this Metal3Machine. If this is set
+but either the `metaData` or `networkData` field is unset, then the
+Metal3Machine controller will wait until those are populated, depending on which
+templates were provided.
+
+The secret containing the metaData or the network data could be provided by the
+user directly.
+
+When CAPM3 controller will set the different fields in the BareMetalHost,
+it will reference the metadata secret and the network data secret
+in the BareMetalHost. If `metaData` or `networkData` field is unset, it will
+also remain unset on the BareMetalHost.
+
+When the Metal3Machine gets deleted, the CAPM3 controller will remove its
+references from the data template object. This will trigger the deletion of the
+secrets generated for this machine.
+
+A new object would be created, a Metal3DataTemplate type.
+
+```yaml
+apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+kind: Metal3DataTemplate
+metadata:
+  name: nodepool-1
+  namespace: default
+  ownerReferences:
+  - apiVersion: infrastructure.cluster.x-k8s.io/v1alpha4
+    controller: true
+    kind: Metal3Cluster
+    name: cluster-1
+spec:
+  metaData: |
+    abc: def
+    local-hostname: {{ machineName }}
+    index: {{ index }}
+  networkData: |
+    {
+        "links": [
+            {
+                "id": "enp1s0",
+                "type": "phy",
+                "ethernet_mac_address": "{{ '{{ bareMetalHostMACByName "eth0" }}' }}"
```

Comment:
I'm having trouble parsing this syntax. It looks like nested {{ }} with one set wrapped in single quotes. Is that right? What does the nesting signify?

Reply:
network data is yaml formatted (see https://github.com/metal3-io/baremetal-operator/blob/master/pkg/provisioner/ironic/ironic.go#L992).